source/logging.c

```diff
@@ -291,6 +291,26 @@ int aws_log_level_to_string(enum aws_log_level log_level, const char **level_str
     return AWS_OP_SUCCESS;
 }
 
+int aws_thread_id_to_string(aws_thread_id thread_id, char *str, size_t length) {
+    AWS_ERROR_PRECONDITION(AWS_THREAD_ID_REPR_LEN == length);
+    AWS_ERROR_PRECONDITION(str && AWS_MEM_IS_WRITABLE(str, length));
+    size_t current_index = 0;
+    unsigned char *bytes = (unsigned char *)&thread_id;
+    for (size_t i = sizeof(aws_thread_id); i != 0; --i) {
+        unsigned char c = bytes[i - 1];
+        int written = snprintf(str + current_index, length - current_index, "%02x", c);
```

Comment:
very debatable: less flexible, but more efficient, to just call snprintf once
```
#if sizeof(aws_thread_id_t) == 4
    snprintf(str, length, "%08x", (uint32_t)thread_id);
#elif sizeof(aws_thread_id_t) == 8
    snprintf(str, length, "%016x", (uint64_t)thread_id);
#else
#    error
#end
```

Reply:
1. TLS limits. Yes, this could be a problem if a platform has limited size. In this case, we would not be able to cache the string repr and we'd have to recompute it each time into a stack-local variable.
2. I think this isn't a concern because the logging thread deals with formatted strings that have already had their thread_id resolved. I.e., the logging thread does not query `tl_current_thread` itself.
3. This would be nice except `sizeof` in macros isn't supported

Reply:
RE: 1
I don't think it should hold up this PR but we should consider adding a thread-local storage system in common that everything else hangs off of so that the CRT's usage is always a single slot.  It could be a mixture of static data (like thread id's stringified value) and key-value data.

Reply:
Thanks Bret. I made your comment into an issue (#565).